# determine LA for GTEx f2 variants and cross-reference with 1000 Genomes superpopulations

# define directories
tmp="/mnt/lab_data/montgomery/nicolerg/tmp"
base="/mnt/lab_data/montgomery/nicolerg/f2-variants"
step1="${base}/1_gtex-admix-all-f2var"
step2="${base}/2_ancestry-per-f2var"
step3="${base}/3_1kg-vcf-intersect"
step4="${base}/4_1kgAF-gtexLA-match"

# provided as an argument to the script call
chr="chr$1"

echo "Working on $chr..."

echo "Step 1: Sort and filter the GTEx f2 variants file" #==================================================

# use files generated by gtex-f2.sh

infile="${step1}/f2variants-${chr}.bed"
sed -i '1s/^/#/' $infile 
bedtools sort -i $infile > "${step1}/sorted-f2variants-${chr}.bed"
infile="${step1}/sorted-f2variants-${chr}.bed"
# remove lines that don't start with ${chr}\t
egrep -v "^$chr" $infile | xargs -Iline sed -i "/line/d" $infile

echo "Step 2: Merge f2 variants with local ancestry" #======================================================

target_region="${step2}/bedintersect.${chr}.out"
echo "#CHROM	START	SNP_ID	REF	ALT	SUBJ1	SUBJ2	POP" > "${target_region}"

while read p
do
	echo $p > "${tmp}/anc.${chr}.line.bed"
	sed -i 's/ /	/g' "${tmp}/anc.${chr}.line.bed"

	ancbed="/mnt/lab_data/montgomery/nicolerg/rfmix-outputs/haplotype-bed/ancestry.${chr}.bed"
	# subjects IDs in fields 7 and 8
	subj1=`cut -f7 "${tmp}/anc.${chr}.line.bed"`
	subj2=`cut -f8 "${tmp}/anc.${chr}.line.bed"`
	# filter BED file
	grep -E "${subj1}|${subj2}" ${ancbed} > "${tmp}/anc.${chr}.filt.bed"

	# run bedtools intersect
	bedtools intersect -a "${tmp}/anc.${chr}.line.bed" -b "${tmp}/anc.${chr}.filt.bed" -wo > "${tmp}/anc.${chr}.bedintersect"

	# check if line count is at least 2
	if [ `wc -l "${tmp}/anc.${chr}.bedintersect" | sed 's/ .*//'` -gt 1 ]; then
		# check if populations are the same
		pop1=`cut -f12 "${tmp}/anc.${chr}.bedintersect" | head -1`
		pop2=`cut -f12 "${tmp}/anc.${chr}.bedintersect" | tail -1`
		if [ "$pop1" = "$pop2" ]; then
			head -1 "${tmp}/anc.${chr}.bedintersect" | cut -f 1,2,4,5,6,7,8,12 >> "${target_region}"
		fi
	fi

	rm "${tmp}/anc.${chr}.line.bed" "${tmp}/anc.${chr}.filt.bed" "${tmp}/anc.${chr}.bedintersect"

done < "${infile}"

gzip "${step1}/f2variants-${chr}.bed"
gzip "${step1}/sorted-f2variants-${chr}.bed"

sed -i "s/^chr//" "${target_region}"

echo "Step 3: Find intersection of f2 variants with 1000 Genomes VCF" #=====================================

vcfin="/mnt/lab_data/montgomery/shared/1KG/ALL.${chr}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
# extract the variants in specified region using bcftools
bcftools view -R ${target_region} -O z -o ${step3}/f2-1kg-intersect-${chr}.vcf.gz ${vcfin}

# grep "AFR" /users/nicolerg/gtex-admix/metadata/1KG-subject-popcodes.tsv | cut -f1 > "${indir}/1kg-AFR.txt"
# grep "ASN" /users/nicolerg/gtex-admix/metadata/1KG-subject-popcodes.tsv | cut -f1 > "${indir}/1kg-ASN.txt"
# grep "EUR" /users/nicolerg/gtex-admix/metadata/1KG-subject-popcodes.tsv | cut -f1 > "${indir}/1kg-EUR.txt"

# vcf-subset -f -e -c ${indir}/1kg-AFR.txt ${outdir}/f2-1kg-intersect-${chr}.vcf.gz | bgzip -c > ${outdir}/f2-1kg-intersect-${chr}-AFR.vcf.gz
# vcf-subset -f -e -c ${indir}/1kg-ASN.txt ${outdir}/f2-1kg-intersect-${chr}.vcf.gz | bgzip -c > ${outdir}/f2-1kg-intersect-${chr}-ASN.vcf.gz
# vcf-subset -f -e -c ${indir}/1kg-EUR.txt ${outdir}/f2-1kg-intersect-${chr}.vcf.gz | bgzip -c > ${outdir}/f2-1kg-intersect-${chr}-EUR.vcf.gz

echo "Step 4: Compute agreement between local ancestry calls and 1000 Genomes populations" #================

python /users/nicolerg/gtex-admix/scripts/f2-analysis/1kg-af-ancestry-crossref.py $step3 $step2 $step4 $chr >> "${step4}/1kg-af-ancestry-crossref.out"

echo
