Scripts for miscellaneous annotations that did not make it into the manuscript. 

## Calculate Fst 
Calculate between- and within- continent Fst. See [`calc_fst.sh`](calc_fst.sh), which also calls [`compile_fst.R`].  

## CADD scores for all SNPs
Download [CADD GRCh38-v1.4](https://cadd.gs.washington.edu/download). Then run [`parse_cadd.py`](parse_cadd.py).  

## Merge annotations  
Run [`merge_all_annotations`](merge_all_annotations) to merge several annotations per eQTL. This requires the following files:
  - [`egenes_master.RData`](https://github.com/nicolerg/gtex-admixture-la/tree/master/eqtl#generate-some-egene-sets-that-are-repeatedly-used-in-downstream-analyses): Lead SNPs for each tested gene in each tissue as generated by [this script](../eqtl/get_uniq_egenes.R)  
  - [`merged-all_no_max.RData`](#calculate-fst): Fst calculations for all SNPs in the 117AX MAF-filtered VCF
  - `EUR_AFR.weir.fst`: Between-continent Fst from [`calc_fst.sh`](calc_fst.sh)  
  - [`all_lead_snp_r2.txt.gz`](https://github.com/nicolerg/gtex-admixture-la/tree/master/annotation#variance-in-snp-genotype-explained-by-local-ancestry): Variance in genotype explained by local ancestry for each lead SNP in `egenes_master.RData`  
  - `gtex.admixed.MAC10.snps.af.txt.gz`: 1000 Genomes allele frequencies for each SNP in the 117AX MAF-filtered VCF  
  - `gtex.admixed.MAC10.all.snps_cadd.tsv`: CADD scores for each SNP in the 117AX MAF-filtered VCF  
