# Annotation

## Calculate variance in gene expression explained by local and global ancestry 
1. Use [`interpolate-tss-la.py`](interpolate-tss-la.py) to extract local ancestry calls for 117AX at annotated hg19 TSSs 
2. Use [`variance_explained_ancestry_gene_expression.R`](variance_explained_ancestry_gene_expression.R) to calculate variance in gene expression explained by each type of ancestry, independent of the other type. This is done using a two-step regression for each expressed gene in each tissue (all genes in GTEx v8 normalized expression files):
  1. First, 117AX gene expression for a single gene is residualized by one ancestry type (local or global)
  2. A second linear regression is fit on the residuals, where the predictors are the opposite ancestry type. Variance explained in the expression of a gene is taken to be the R-squared value from the second fit.  

## Calculate LD between pairs of lead SNPs that are different between LocalAA and GlobalAA  
1. Run [`pairs_for_ld.R`](pairs_for_ld.R) to generate a file `evariant_pairs_test_ld.txt` which includes all of the pairs of SNPs for which to calculate LD  
2. Prepare input files for PLINK: 
    ```bash 
    cut -f 3,4 evariant_pairs_test_ld.txt | sed "s/ / /g" > pairs_in.txt
    split -d -n l/15 pairs_in.txt pairs_in_ 
    ```
3. Run [`plink_ld.sh`](plink_ld.sh) on each `pairs_in_*` chunk to calculate LD between pairs of SNPs using PLINK: `for file in pairs_in_*; do bash plink_ld.sh ${file} & done`
4. Concatenate results  
5. Run [`define_diff_lead_by_ld.R`](define_diff_lead_by_ld.R) to restrict the definition of "different lead SNPs" to those with LD < 1  

## Calculate Fst (between- and within- continent)

## CADD scores for all SNPs

## EUR-vs-AFR Fst 

## Allele frequencies from 1000 Genomes

## Variance in SNP genotype explained by local ancestry 
Use [`r2_all_snps.R`](r2_all_snps.R) to get variance in genotype explained by local ancestry for all lead SNPs in the LocalAA/GlobalAA eQTL analysis (117AX only).  

Then use [`gtex_variant_la_r2_by_tissue.R`](gtex_variant_la_r2_by_tissue.R) to calculate variance in genotype explained by local ancestry for all GTEx v8 eSNPs (all subjects).  

## Merge annotations  
Run [`merge_all_annotations`](merge_all_annotations) to merge several annotations per eQTL. This requires the following files:
  - [`egenes_master.RData`:](https://github.com/nicolerg/gtex-admixture-la/tree/master/eqtl#generate-some-egene-sets-that-are-repeatedly-used-in-downstream-analyses) Lead SNPs for each tested gene in each tissue as generated by [this script](../eqtl/get_uniq_egenes.R)  
  - [`merged-all_no_max.RData`:](https://github.com/nicolerg/gtex-admixture-la/tree/master/annotation#calculate-fst-between--and-within--continent) Fst calculations for all SNPs in the 117AX MAF-filtered VCF  
  - `EUR_AFR.weir.fst`:
  - [`all_lead_snp_r2.txt.gz`:](https://github.com/nicolerg/gtex-admixture-la/tree/master/annotation#variance-in-snp-genotype-explained-by-local-ancestry) Variance in genotype explained by local ancestry for each lead SNP in `egenes_master.RData`  
  - `gtex.admixed.MAC10.snps.af.txt.gz`: 1000 Genomes allele frequencies for each SNP in the 117AX MAF-filtered VCF  
  - `gtex.admixed.MAC10.all.snps_cadd.tsv`: CADD scores for each SNP in the 117AX MAF-filtered VCF  
